pkg/storage/replicate_queue.go

```diff
@@ -934,11 +934,10 @@ func (rq *replicateQueue) findTargetAndTransferLease(
 ) (bool, error) {
 	// Learner replicas aren't allowed to become the leaseholder or raft leader,
 	// so only consider the `Voters` replicas.
-	candidates := filterBehindReplicas(repl.RaftStatus(), desc.Replicas().Voters())
 	target := rq.allocator.TransferLeaseTarget(
 		ctx,
 		zone,
-		candidates,
+		desc.Replicas().Voters(),
```

Comment:
I agree that this particular check is misguided, but there'll probably be some way in which we're shooting ourselves in the foot if we just remove it. For example, I'm not seeing any code in `TransferLeaseTarget` that would stop it from transferring to a node that just went down, so the transferred lease will time out, someone else will get the lease, attempt to transfer again, time out, ... until the store dead timeout has passed. I think we at least want to check the follower's liveness. There is some kind of check here:

https://github.com/cockroachdb/cockroach/blob/0cf73600f5dce7d88666472274ed40d2ad44f13b/pkg/storage/allocator.go#L852

but note that it's never reached if we have lease preferences, and it uses the 5min timeout.

`store.cfg.NodeLiveness.GetLivenessStatusMap()` (or the per-node `.GetLiveness(nodeID).IsLive` seem like a better signal.

So you could replace `filterBehindReplicas` with a new `onlyLiveReplicas(...)`, and perhaps also remove some of the inefficient liveness checking in `TransferLeaseTarget` (though all callers would have to prune, maybe it's better to push the liveness map into TransferLeaseTarget instead?)

Reply:
This is already the case. The first thing that `TransferLeaseTarget` does is call `StorePool.getStoreList`. It grabs the `NodeLiveness` record for each node and calls `Liveness.LivenessStatus` and in turn `Liveness.IsLive`.